### ISSUE TITLE
Runtime error when converting T5 decoder with multiple dynamic axes inputs

### ISSUE THREAD

**Issue by amanpreet692** [microsoft/onnxruntime#5646]
**Describe the bug**
I have been trying to convert a T5 model. The encoder works as expected but decoder expects two inputs:
i) encoder states (Works fine)
ii) past states from each layer(This is a list(12)->tuple(4)->Tensor collection which is converted into 48 separate inputs.  A similar structure is output as well which works fine but input is causing the problem.

On running inference, get the following error:
```python
[E:onnxruntime:, sequential_executor.cc:318 Execute] Non-zero status code returned while running Transpose node.
Name:'Transpose_207' Status Message: /Users/runner/work/1/s/onnxruntime/core/framework/op_kernel.cc:62 
OrtValue *onnxruntime::OpKernelContext::OutputMLValue(int, const onnxruntime::TensorShape &, size_t) status.IsOK() was false. 
Shape mismatch attempting to re-use buffer. {1,12,64,2} != {1,12,64,933}. 
Validate usage of dim_value (values should be > 0) and dim_param (all values with the same string should equate 
to the same size) in shapes in the model.
```
 Model definition:
```python
    input_ids = torch.tensor([[42] * 10])
    past_state_input_pre = torch.rand((1,12,1,64))
    past_state_input_post = torch.rand((1, 12, 10, 64))
    past_key_value_states = [(past_state_input_pre, past_state_input_pre, past_state_input_post, past_state_input_post) for i in range(12)]

    past_val_outputs = {'past_states_op_'+str(i): {0:'batch', 2: 'sequence'} for i in range(48)}
    past_val_inputs = {'past_states_ip' + str(i): {0: 'batch', 2: 'sequence'} for i in range(48)}
    dynamix_axes_dict = {
                              'input_ids': {0:'batch', 1: 'sequence'},
                              'encoder_hidden_states': {0:'batch', 1: 'sequence'}
                            }
    dynamix_axes_dict.update(past_val_inputs)
    dynamix_axes_dict.update({'logits': {0:'batch', 1: 'sequence'}})
    dynamix_axes_dict.update(past_val_outputs)
    output_names_list = ['logits'] + ['past_states_op_' + str(i) for i in range(48)]
    input_names_list = ['input_ids', 'encoder_hidden_states'] + ['past_states_ip' + str(i) for i in range(48)]
    # Exports to ONNX
    _ = torch.onnx.export(
                            decoder_with_lm_head,
                            (torch.tensor([[42]]), simplified_encoder(input_ids), past_key_value_states),
                                   f"{output_prefix}-decoder-with-lm-head.onnx",
                                   export_params=True,
                            opset_version=12,
                            input_names=input_names_list,
                            output_names=output_names_list,
                            dynamic_axes= dynamix_axes_dict, do_constant_folding=False, verbose=True)

```

Please let me know if I am missing anything. Sharing the model file as 

**Urgency**
None

**System information**
- OS Platform and Distribution Mac 10.15.7:
- ONNX Runtime installed from (source or binary): pip
- ONNX Runtime version: 1.4.0
- Python version: 3.7
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: >10 (inferring on CPU though)
- GPU model and memory: N/A

**To Reproduce**
- Describe steps/code to reproduce the behavior.
   Passing the input while running inference:
   i) 2-d tensor with 1 element eg. [[42]]
  ii) 3-d tensor with shape (1,sequence, 768)
  iii) past_state_input_pre = torch.rand((1,12,1,64))
      past_state_input_post = torch.rand((1, 12, 10, 64))
      48 tensors using the above with a sequence of 2 past_state_input_pre, 2 past_state_input_post 

- Attach the ONNX model to the issue (where applicable) to expedite investigation.
[Model](https://drive.google.com/file/d/184EtnezrNzEFCi02BljWPEXbxGZUD_xo/view?usp=sharing)
**Expected behavior**
Should get back 2 outputs: logits of shape (batch, sequence, 32168), 48 tensors with shapes as above

**Screenshots**

**Additional context**
Model:  [T5- HuggingFace](https://huggingface.co/t5-base)

**Comment by yufenglee** [comment]
@wangyems, @tianleiwu , please help to take a look.

**Comment by amanpreet692** [comment]
Thank you for picking it up...plz post here if more info is needed from my end!

**Comment by amanpreet692** [comment]
Guys, any updates? We have been stuck on this for a while now.

**Comment by wangyems** [comment]
Hi @amanpreet692, sorry for the late response. I just took a quick look at your attached model and I didn't see the problematic node "Transpose_207" in that model which cause the runtime error. Could you check if the uploaded model is the one you ran inference upon?

**Comment by amanpreet692** [comment]
Hey @wangyems thanks for getting back.
With the uploaded model I get the following error:
```python 
onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero 

status code returned while running Transpose node. Name:'Transpose_216' Status Message: 

/Users/runner/work/1/s/onnxruntime/core/framework/op_kernel.cc:62 OrtValue 

*onnxruntime::OpKernelContext::OutputMLValue(int, const onnxruntime::TensorShape &, size_t) status.IsOK() was false. 

Shape mismatch attempting to re-use buffer. {5,12,64,2} != {5,12,64,933}. Validate usage of dim_value (values should be > 0) 

and dim_param (all values with the same string should equate to the same size) in shapes in the model.

```

Transpose_216 should be present in the model. Hope this helps!

**Comment by wangyems** [comment]
Hi @amanpreet692, the transpose node looks good to me as it requires the correct shape from the input. How is your ort_inputs organized? Could you provide the part of code that you run inference so that I can investigate further?

**Comment by amanpreet692** [comment]
Hey @wangyems, thanks for checking that!
This is the inference code:
```python 
past_state_list = []
[past_state_list.extend(list(tup)) for tup in past_key_value_states] #past_key_value_states is the input here
past_state_list = [tensor.numpy() for tensor in past_state_list]
input_dict = {"input_ids": input_ids[:, -1:].cpu().numpy(), "encoder_hidden_states": 
encoder_hidden_states.cpu().numpy()}
input_names = decoder_sess.get_inputs()   #decoder_sess is the InferenceSession object
input_dict.update({input_names[i+2].name:past_state_list[i] for i in range(len(past_state_list))})
decoder_outputs = decoder_sess.run(None, input_dict)

```
You can use the following pickle file as sample input for the past_key_value_states:
[sample past_key_value_states](https://drive.google.com/file/d/1FfuAk83UbuDmHTsc_6__wv-FhmkI2fIY/view?usp=sharing)

Let me know if anything else is needed.

**Comment by wangyems** [comment]
Thank you for providing the code @amanpreet692. After looking at the way you preparing the input, I think the cause is that the input shape is not align with what onnx model requires. e.g. There are a couple of inputs that require float32[batch, 12, sequence, 64], you need to make sure that the actual inputs share the same batch number and sequence number. This seems not a framework issue.

**Comment by amanpreet692** [comment]
Hey @wangyems Thanks again for checking.
Can you please elaborate a bit on your investigation? coz I am still confused as to what's wrong here.
If you @ the input shapes in the errors, {5,12,64,2} != {5,12,64,933}, here when passing the inputs the shapes are actually {5,12,2,64} and {5,12,933,64}. My assumption was that due to the transpose operation the shapes are re-arranged and as mentioned in the error a buffer with the previously used shape is attempted to be re-used which is causing the error.
Is that not so?

**Comment by wangyems** [comment]
Hi @amanpreet692, According to your code given [here](https://github.com/microsoft/onnxruntime/issues/5646#issue-733116902), the inputs do not have the same sequence value, they are sometimes 1 and others are 10. If you fix this, the buffer reuse issue may still occur but it's non-blocking. It may be due to that your model have some initializers serving as both input and output. Please check if your model looks as you expected.

**Comment by amanpreet692** [comment]
Ok, but that is the reason the sequence values have been declared in dynamic axis because the sequence shape is not fixed. I am afraid I am nt aware of any other way to prevent/fix this. It would be great if you could point me in the right direction. The same thing works while getting the output btw.
Thanks

**Comment by amanpreet692** [comment]
Hey @wangyems, I was able to get this to work today by removing the 48 numpy arrays from dynamic axes along with their output names as it didn't seem to matter if I mentioned dynamic axes for the output or not.
But every 3rd and 4th of those output nodes were being used again as input nodes(presumably due to same values) and their shape was not being considered as dynamic. This is all a theory which might actually point to a bug in the exporting code where the output shape doesn't matter for dynamic axes.
It would be great if you could comment on this!
Thanks

**Comment by wangyems** [comment]
Hi @amanpreet692, great to hear that you made it work. The concept of dynamic axis is like a variable - any inputs shapes that use this variable will share the same value during inferencing. So if those values are not the same then the dynamic variable shall not be considered. Regarding your export question, you may want to take a look at https://pytorch.org/docs/stable/onnx.html and see if it answers your question.
